plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -2327,7 +2360,8 @@ else if (isMetadataColumnId(columnHandle.getId())) {
         }
 
         if (newEnforcedConstraint.equals(table.getEnforcedPredicate())
-                && newUnenforcedConstraint.equals(table.getUnenforcedPredicate())) {
+                && newUnenforcedConstraint.equals(table.getUnenforcedPredicate())
+                && constraint.getPredicateColumns().isEmpty()) {
```

Comment:
Sorry I'm not sure I understand the problem. Why was it necessary to add the extra condition to the if statement?

Reply:
Purely from the semantics of the `ConnectorMetadata#applyFilter` interface, if calling this method has no effect on the call, it is very important for the connector to return `Optional.empty()` to prevent the optimizer from entering an infinite loop, `constraint.getPredicateColumns().isEmpty( )` is true means we don't need to push down predicateColumns to iceberg.

Reply:
I'd think you need a similar check for `predicatedColumns` as the other fields in this condition, check that it is different from the existing one in the IcebergTableHandle

Reply:
Similarly to the Hive version here: https://github.com/trinodb/trino/blob/master/plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java#L2960C1-L2961C1

Reply:
> This parameter will be set to true only when AddExchanges#visitFilter. And AddExchanges in PlanOptimizers is not added as a set of rules in IterativeOptimizer

@findepi @zhangminglei @marton-bod @alexjo2144 @Praveen2112 
This is not true since https://github.com/trinodb/trino/pull/16019/files#diff-624ea155d75175ba8802d5f93af759bdb88c6f63f1ca9660d6704892e1033f1dR814-R825
AFAIU, we don't see an infinite loop only because of `io.trino.sql.planner.iterative.rule.PushPredicateIntoTableScan#arePlansSame`. 
If in the future `io.trino.plugin.iceberg.IcebergTableHandle#equals` will return `false` when comparing 2 tables of 2 different iterations, then we'll see an infinite loop.

Reply:
@assaf2  Thanks for bringing this up.
 Looking at the `io.trino.spi.connector.ConnectorTableHandle` I don't see a mention of `equals`, so I would assume the connector does not have to implement that method. If this is a wrong assumption, maybe it is enough to fix the interface.

@assaf2 can you please convert this to an issue so that we can have Martin involved more conveniently?

Reply:
https://github.com/trinodb/trino/issues/19876